docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+    it considers it is caught up, can now be recovered from past blocks.
+
+    This solution has two main drawbacks.
+
+    - It changes the blockchain format, in particular, the way light blocks are verified; thus breaking
+      backwards compatiblity in features such as light clients and IBC (the latter relying on the former).
+    - The extra information (i.e., the vote extensions) that is now kept in the blockchain is not really
+      needed *at every height* for a late node to catch up. This information is only needed to be able
+      to *propose* at the height the validator considers itself as caught-up. If a validator is indeed
+      late for height *h*, it is useless (although correct) for it to call `PrepareProposal`, or
+      `ExtendVote`, since the block is already decided.
+
+- **2.** *Skip* propose *step in Tendermint algorithm*.
+
+    This solution consists in modifying the Tendermint algorithm to skip the *send proposal* step in
+    heights where the node does not have the required vote extensions to populate the call to
+    `PrepareProposal`. The assumption is that the validator is late and, therefore, up-to-date
+    validators have already proposed (and decided) for that height.
+    A small variation of this solution is, rather than skipping the *send proposal* step, the validator
+    sends a special *empty* or *bottom* proposal to signal other nodes that it is not ready to propose
+    at (any round of) the current height.
+
+    The appeal of this solution is its simplicity. A possible implementation does not need to extend
+    the data structures, or change the current catch-up mechanisms implemented in the blocksync or
+    in the consensus reactor. When we lack the needed information (vote extensions), we simply rely
+    on another correct validator to propose a valid block in other rounds of the current height.
+
+    However, this solution can be attacked by a byzantine node in the network in the following way.
+    Let us consider the following scenario:
+
+    - all validators send out precommit messages, with vote extensions, for height *h*, round *r*,
+      roughly at the same time,
+    - all validators then wait until they gather enough precommit messages for height *h*, round *r*,
+      in order to decide in height *h*,
+    - all those precommit messages (and any further message for height *h*) get delayed indefinitely,
+    - an intermediate (malicious) node *m* manages to gather more than *2n/3* of the precommit messages
+      for height *h*, round *r*, as well as the proposed block the precommit messages refer to,
+    - node *m* uses the precommit messages to build a commit (not extended commit), which it then sends
+      to *all* validators in order to convince them they are late,
+    - all validators receive the *catch-up* message from *m*, decide on height *h*, and proceed to
+      height *h+1*.
```

Comment:
This is very nicely described. I'm not sure I understand the 'catch-up' message here. Is this a message that exists today within Tendermint?

I think that a slight variation of this is that a malicious node, or set of nodes could get a recovering node to blocksync to the head of the chain and not provide those nodes with vote extensions. This would force the nodes to all panic at the head of the chain and be left in an unrecoverable state.

Reply:
Yes, but I wanted to leave it abstract... maybe I should reconsider.
The catch-up message, in current Tendermint implementation would be in two forms:
* A block received in blocksync
* A set of precommit votes extracted from the commit sent to the late node

Reply:
Here, I was trying to stay abstract and not consider our implementation, if possible. But after your comments, maybe it's just not possible to do that, while having a clear text.

I'll go through this solution again and will add some details on reactors.

Reply:
This is on the reactor (gossip), not in the actual implementation. The commit is loaded from the block store and sent do a peer that is in a previous height.

Reply:
Right, but, as far as I know, that doesn't actually signal to the node that it is not at the head of the chain. The node "thinks" that it is participating in an active consensus round.

Reply:
Let me insist that I wrote this section in abstract terms, aimed at readers that are familiar with consensus and the Tendermint (white paper) algorithm, but not necessarily acquainted with the details of our current code base.
We have the "Possible Implementations" later on, where we can better discuss details of the current code